SOLR-15277: Schema designer UI and supporting backend

thelabdude opened the pull request:
# Description

See https://issues.apache.org/jira/browse/SOLR-15277

# Solution

Introduce a new Schema Designer section of the Admin UI, see screenshots in JIRA.

# Tests

Unit tests added for backend classes.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

thelabdude commented on solr/webapp/web/js/angular/services.js:
```diff
@@ -254,6 +254,17 @@ solrAdminServices.factory('System',
+       put: {method: "PUT"},
+       postXml: {headers: {'Content-type': 'text/xml'}, method: "POST"},
+       postCsv: {headers: {'Content-type': 'application/csv'}, method: "POST"},
+       upload: {method: "POST", transformRequest: angular.identity, headers: {'Content-Type': undefined} }
```
Hi @MarcusSorealheis, thanks for taking a look ...

With jQuery, the default content-type is always `application/json` except where I've declared it explicitly. Almost all of the API calls the designer makes to the backend send and expect JSON.

MarcusSorealheis replied:
That makes sense. I am still testing out the tool and profiling the network requests. I had not gotten to this request in the real world, though it was on my list. This was my reaction when I was simply reading the code.

Awesome job. Important feature.

MarcusSorealheis commented on solr/core/src/java/org/apache/solr/core/CoreContainer.java:
```diff
@@ -805,6 +806,12 @@ public void load() {
+      SchemaDesignerAPI schemaDesignerAPI = new SchemaDesignerAPI(this);
+      containerHandlers.getApiBag().registerObject(schemaDesignerAPI);
+    } // Schema Designer not available in standalone (non-cloud) mode
+
```
Do you think the ideal programmatic experience would be for users in Standalone mode to receive this information (`not available in Standalone mode`) via API response?

thelabdude replied:
Not sure? The API is mostly intended to support the UI ... this feature needs UI and I don't foresee too many people wanting to interact with the Schema Designer API directly w/o using the Web Admin UI. The Schema Designer link in the Admin UI should not even appear in standalone mode.

MarcusSorealheis replied:
ahh. That makes sense. I wasn't sure if there would be two different personas interfacing, e.g, someone posting to a shared instance a schema they have ben toying with locally and another person reviewing the same schema.

HoustonPutman commented:
Some thoughts while playing around:

- [ ] I would like to be able to populate the example document with a document that exists. Maybe a button in the query results.
- [ ] docValues should be enabled by default when creating a fieldType/field
- [ ] Do we recommend the use of `useDocValuesAsStored` instead of `stored`?
- [ ] When you analyze the fields it sometimes brings up a question about language and some options. Not clear that this is talking about the entire schema. Would be better to highlight the configName in the menu.
- [ ] Maybe guess long strings as text fields, not string fields?
- [ ] Would be nice to be able to filter on a field not having a feature (docValues)
- [ ] When filtering fields, it would be nice to filter the dynamic fields as well.
- [ ] In the future it would be nice to have information on each of the options. Not sure what the best way to go about this is though.

epugh commented on solr/webapp/web/partials/schema-designer.html:
```diff
@@ -0,0 +1,919 @@
+      </div>
+
+      <div id="confirm-dialog" class="action add" data-rel="add" ng-show="showConfirmEditSchema" style="display:block;">
+        <p class="clearfix warn"><span>Warning: You've chosen to load an existing schema that is already being used by active collections. Making changes to the '{{confirmSchema}}' schema will impact existing documents in these collections. Please proceed with caution.<br/><br/></span></p>
```
"these collections" or should it be "those collections".   nitpick I know!

epugh replied:
actually, since you do list the collections, maybe these is right word!

epugh commented:
1. when "Analyzing your sample data, schema will load momentarily" text comes up, could we swap the gear for a moving animated gear icon?  

1. download products a file called "download", not a "download.zip", or even better...  "techproducts_configset.zip"

1. Is the config a config?  Or a ConfigSet????

1. I wish the folder icon when clicked on would open it.  Versus the small > arrow.  This I find annoying in the SolrCloud ZK tree as well, though there you might argue you want to look at metadata..   I wish there it did both showing metadata and opened, and on schema designer it opened!

1. I wonder if you could get more screen space by having any kind of paning close/open options?   

1. in the facet listing, when you shift click you get all between the first and the second click, versus shift click letting you pick them individually, which I think is what you wanted!

1. are there any features that you expose that NO ONE should ever use and could be removed???

thelabdude commented:
Thanks for taking a look @HoustonPutman ... I addressed most of your concerns but there's some nuance around a few of your requests:

>  I would like to be able to populate the example document with a document that exists. Maybe a button in the query results.

I haven't implemented this yet. The query results only include sample documents already, so maybe you want to revise the sample documents in the paste text area? I think that's a valid thing to do and will think about this some more. My only concern would be adding another option that may complicate the experience. Can you elaborate on the user experience you want here?

> Do we recommend the use of useDocValuesAsStored instead of stored?

I'm not sure about whether that is a recommendation? Seems reasonable to me that if you're already using DocValues, then `useDocValuesAsStored` seems more efficient from a storage perspective (if you store and use doc values, then seems like you're doing double storage), but maybe there is some performance hit for doing this? I can certainly turn that off if we don't like defaulting to stored=false and useDocValuesAsStored=true.

> When you analyze the fields it sometimes brings up a question about language and some options. Not clear that this is talking about the entire schema. Would be better to highlight the configName in the menu.

There are some gremlins in that jstree we're using where the selected node doesn't always show as selected in the tree. I've tried to improve this with my latest commit and seems to be rendering correctly. When the root node is selected, you see the panel with languages and options. Those apply to the entire schema. I added the schema name to that panel to help make this clear.

> In the future it would be nice to have information on each of the options. Not sure what the best way to go about this is though.

I think some of the options are pretty confusing and need supporting documentation, so have deferred to the ref guide via the help icon links vs. adding more docs directly in the Schema Designer UI.

thelabdude commented:
Thanks @epugh ... my latest commit should address most of your concerns ...

> when "Analyzing your sample data, schema will load momentarily" text comes up, could we swap the gear for a moving animated gear icon?

done.

> download products a file called "download", not a "download.zip", or even better... "techproducts_configset.zip"

done.

> Is the config a config? Or a ConfigSet????

Where in the UI are you seeing mention of config? I've tried to refer to `Configset` but that's not really a term new users would really understand anyway, so I've taken some liberties on just focusing on `schema` in most places.

> I wish the folder icon when clicked on would open it. Versus the small > arrow. This I find annoying in the SolrCloud ZK tree as well, though there you might argue you want to look at metadata.. I wish there it did both showing metadata and opened, and on schema designer it opened!

Double-click works, but I couldn't figure out how to get it to open on single-click. May not be supported by the underlying `jstree` UI component we're using.

> I wonder if you could get more screen space by having any kind of paning close/open options?

We could probably allow closing the Sample documents area once the data is loaded, to give more room for the Schema Editor (middle section). There's also some wasted space on the right for Text Analysis if you're not looking at text fields. I'm curious what the minimum resolution / screen size we should design this for though? Having a view of your sample docs, schema, text analysis, and query tester form / results all in once place is crucial to the design.

> in the facet listing, when you shift click you get all between the first and the second click, versus shift click letting you pick them individually, which I think is what you wanted!

Ok, I removed the tip about using shift as I'm not sure how it works on Windows. On Mac, it's cmd + click to select non-adjacent items. I guess we can just assume people will figure it out on their OS.

> are there any features that you expose that NO ONE should ever use and could be removed???

Not sure? You tell me ;-) Solr schema design is a complicated process and I've designed a solution to help with that, but the fact remains it's still pretty complicated but at least we can give users quick feedback and let them iterate in a safe manner.

HoustonPutman commented:
> > I would like to be able to populate the example document with a document that exists. Maybe a button in the query results. 

> I haven't implemented this yet. The query results only include sample documents already, so maybe you want to revise the sample documents in the paste text area? I think that's a valid thing to do and will think about this some more. My only concern would be adding another option that may complicate the experience. Can you elaborate on the user experience you want here?

For sure. revising the sample documents is exactly what I want. If I manually input a complex document, I don't want to have to do it again by scratch. This could be in the next iteration, it's definitely not a deal breaker, just an inconvenience. Maybe there's just an option to see the JSON in the sample results screen, so I could copy and paste.

> > Do we recommend the use of useDocValuesAsStored instead of stored?

> I'm not sure about whether that is a recommendation? Seems reasonable to me that if you're already using DocValues, then useDocValuesAsStored seems more efficient from a storage perspective (if you store and use doc values, then seems like you're doing double storage), but maybe there is some performance hit for doing this? I can certainly turn that off if we don't like defaulting to stored=false and useDocValuesAsStored=true.

I think I'm in the wrong here. I remember hearing rumblings of certain features that didn't support fields with `useDocValuesAsStored` only. But it seems like it's the default option in schemas with version 1.6. Anyway, if its an issue we can certainly fix it, or the missing `useDocValuesAsStored` features, in the future.

> > In the future it would be nice to have information on each of the options. Not sure what the best way to go about this is though.

> I think some of the options are pretty confusing and need supporting documentation, so have deferred to the ref guide via the help icon links vs. adding more docs directly in the Schema Designer UI.

👍 

Great work as always!

thelabdude commented:
> Maybe there's just an option to see the JSON in the sample results screen, so I could copy and paste.
Ha! Great idea ;-) I've been wanting to add that anyway, so this is another reason to do so. Thanks for the suggestion.

MarcusSorealheis commented:
Schema Designer should not simply throw generic errors that do no help the user. For instance, when a user tries to create a field that already exists, there is just an error processing banner. Ideally, you could grab the important detail in the response body and show it to the user. Consider `data["error"]["details"][0]["errorMessages"][0]` as seen below to populate the banner with something helpful to the user:
```
Possibly unhandled rejection: {"data":{"responseHeader":{"status":400,"QTime":1},"error":{"metadata":["error-class","org.apache.solr.api.ApiBag$ExceptionWithErrObject","root-error-class","org.apache.solr.api.ApiBag$ExceptionWithErrObject"],"details":[{"add-field":{"stored":"true","indexed":"true","uninvertible":"true","name":"name","type":"string"},"errorMessages":["Field 'name' already exists.\n"]}],"msg":"error processing commands","code":400}},"status":400,"config":{"method":"POST","transformRequest":[null],"transformResponse":[null],"jsonpCallbackParam":"callback","data":{"add-field":{"stored":"true","indexed":"true","uninvertible":"true","name":"name","type":"string"}},"url":"techproducts/schema","params":{"wt":"json","_":1622686070279},"headers":{"Accept":"application/json, text/plain, */*","X-Requested-With":"XMLHttpRequest","Content-Type":"application/json;charset=utf-8"},"timeout":10000},"statusText":"Bad Request","xhrStatus":"complete","resource":{"add-field":"..."}}```

MarcusSorealheis commented:
there's also a few security exceptions that if thrown in the wrong corporate org could start problematic conversations running SOLR. Everything is under scrutiny right now given the recent events.

To fix those errors referencing image resources being blocked by the CSP, we need to add this clause to `jetty.xml`: `img-src 'self' data:`. The change will still restrict img resources to the host but it will allow SVG and IMG. We cannot be more specific because of the port. :) It is a replacement of the existing `img-srv 'self' data`.

MarcusSorealheis reviewed (changes requested):
I think once you fix the issues I have pointed out, it feels good as an MVP of the schema designer's return. Also, the jetty fix may be impacting many SOLR users at the moment. Maybe the person who added it has some rationale for why they did? Although, I don't perceive the `data:` directive as risky when applied to `img-src 'self'`.

thelabdude commented:
> Schema Designer should not simply throw generic errors that do no help the user. For instance, when a user tries to create a field that already exists, there is just an error processing banner. Ideally, you could grab the important detail in the response body and show it to the user. Consider `data["error"]["details"][0]["errorMessages"][0]` as seen below to populate the banner with something helpful to the user:
> 
> ```
> Possibly unhandled rejection: {"data":{"responseHeader":{"status":400,"QTime":1},"error":{"metadata":["error-class","org.apache.solr.api.ApiBag$ExceptionWithErrObject","root-error-class","org.apache.solr.api.ApiBag$ExceptionWithErrObject"],"details":[{"add-field":{"stored":"true","indexed":"true","uninvertible":"true","name":"name","type":"string"},"errorMessages":["Field 'name' already exists.\n"]}],"msg":"error processing commands","code":400}},"status":400,"config":{"method":"POST","transformRequest":[null],"transformResponse":[null],"jsonpCallbackParam":"callback","data":{"add-field":{"stored":"true","indexed":"true","uninvertible":"true","name":"name","type":"string"}},"url":"techproducts/schema","params":{"wt":"json","_":1622686070279},"headers":{"Accept":"application/json, text/plain, */*","X-Requested-With":"XMLHttpRequest","Content-Type":"application/json;charset=utf-8"},"timeout":10000},"statusText":"Bad Request","xhrStatus":"complete","resource":{"add-field":"..."}}```
> ```

Can you please provide reproduction steps for how you got this error (including browser details)
<img width="357" alt="Screen Shot 2021-06-03 at 9 32 19 AM" src="https://user-images.githubusercontent.com/417074/120671957-1c3aad00-c44f-11eb-8c62-ff6051ba9316.png">
? Schema Designer UI should be displaying user friendly errors in a dialog. Moreover, I don't even know how you got this error since the JS code validates the field doesn't already exist before submitting the request. See screenshot of what I see in my env.

thelabdude commented:
> there's also a few security exceptions that if thrown in the wrong corporate org could start problematic conversations running SOLR. Everything is under scrutiny right now given the recent events.
> 
> To fix those errors referencing image resources being blocked by the CSP, we need to add this clause to `jetty.xml`: `img-src 'self' data:`. The change will still restrict img resources to the host but it will allow SVG and IMG. We cannot be more specific because of the port. :) It is a replacement of the existing `img-srv 'self'`.

This doesn't sound like it is Schema Designer UI specific. I believe I'm referencing icons the same way all the other screens do. If you think this is a Schema Designer specific issue, please provide a link to the offending code. Otherwise, please open another JIRA to address that problem independently of this issue.

thelabdude commented on solr/webapp/web/index.html:
```diff
@@ -26,6 +26,7 @@
+  <link rel="stylesheet" type="text/css" href="css/angular/schema-designer.css?_=${version}">
   <link rel="stylesheet" type="text/css" href="css/angular/cloud.css?_=${version}">
   <link rel="stylesheet" type="text/css" href="css/angular/cores.css?_=${version}">
   <link rel="stylesheet" type="text/css" href="css/angular/collections.css?_=${version}">
```
hmmm ... This too seems unrelated to the Schema Designer UI. Please raise another JIRA.

MarcusSorealheis commented:
> > there's also a few security exceptions that if thrown in the wrong corporate org could start problematic conversations running SOLR. Everything is under scrutiny right now given the recent events.
> > To fix those errors referencing image resources being blocked by the CSP, we need to add this clause to `jetty.xml`: `img-src 'self' data:`. The change will still restrict img resources to the host but it will allow SVG and IMG. We cannot be more specific because of the port. :) It is a replacement of the existing `img-srv 'self'`.
> 
> This doesn't sound like it is Schema Designer UI specific. I believe I'm referencing icons the same way all the other screens do. If you think this is a Schema Designer specific issue, please provide a link to the offending code. Otherwise, please open another JIRA to address that problem independently of this issue.

It's not Schema designer specific. It's Solr. I'll open a JIRA, thanks :)

MarcusSorealheis commented:
> > Schema Designer should not simply throw generic errors that do no help the user. For instance, when a user tries to create a field that already exists, there is just an error processing banner. Ideally, you could grab the important detail in the response body and show it to the user. Consider `data["error"]["details"][0]["errorMessages"][0]` as seen below to populate the banner with something helpful to the user:
> > ```
> > Possibly unhandled rejection: {"data":{"responseHeader":{"status":400,"QTime":1},"error":{"metadata":["error-class","org.apache.solr.api.ApiBag$ExceptionWithErrObject","root-error-class","org.apache.solr.api.ApiBag$ExceptionWithErrObject"],"details":[{"add-field":{"stored":"true","indexed":"true","uninvertible":"true","name":"name","type":"string"},"errorMessages":["Field 'name' already exists.\n"]}],"msg":"error processing commands","code":400}},"status":400,"config":{"method":"POST","transformRequest":[null],"transformResponse":[null],"jsonpCallbackParam":"callback","data":{"add-field":{"stored":"true","indexed":"true","uninvertible":"true","name":"name","type":"string"}},"url":"techproducts/schema","params":{"wt":"json","_":1622686070279},"headers":{"Accept":"application/json, text/plain, */*","X-Requested-With":"XMLHttpRequest","Content-Type":"application/json;charset=utf-8"},"timeout":10000},"statusText":"Bad Request","xhrStatus":"complete","resource":{"add-field":"..."}}```
> > ```
> 
> Can you please provide reproduction steps for how you got this error (including browser details)? Schema Designer UI should be displaying user friendly errors in a dialog. Moreover, I don't even know how you got this error since the JS code validates the field doesn't already exist before submitting the request. See screenshot of what I see in my env. <img alt="Screen Shot 2021-06-03 at 9 32 19 AM" width="357" src="https://user-images.githubusercontent.com/417074/120671957-1c3aad00-c44f-11eb-8c62-ff6051ba9316.png">

1. pull down the repo at this point
2. `gh checkout pr 42` 
3. ./gradlew assemble
4. cd into the solr snapshot
5. bin/solr start -e techproducts
6. open Chrome Version 91.0.4472.77 (Official Build) (x86_64)
7. click schema
8. add field
9. see screenshot - type name in the field and set it to the string type
10. see screenshot - click add field
![image](https://user-images.githubusercontent.com/2353608/120701247-76e00300-c467-11eb-8af2-a3b5f59baeeb.png)
![image](https://user-images.githubusercontent.com/2353608/120701329-91b27780-c467-11eb-9b34-4f6381d98378.png)

MarcusSorealheis commented:
I just rebuilt from nothing and got the same error

thelabdude commented:
Schema Designer only runs in cloud-mode, from your screenshot, you're running in standalone. See screenshots of the Schema Designer UI in the JIRA

MarcusSorealheis commented:
Ignore this.. Looking into the client side errors now. First two nits and SOLR-wide issue persists. New JIRA incoming.

MarcusSorealheis commented:
> This doesn't sound like it is Schema Designer UI specific. I believe I'm referencing icons the same way all the other screens do. If you think this is a Schema Designer specific issue, please provide a link to the offending code. Otherwise, please open another JIRA to address that problem independently of this issue.

I added it here [#164](https://github.com/apache/solr/pull/164)
